Assets/Tests/Runtime/ConnectionManagementTests.cs

```diff
@@ -0,0 +1,593 @@
+            m_ClientConnectionStateSequences = new List<ConnectionState>[NumberOfClients];
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                var clientId = i;
+                m_ClientConnectionStateSequences[clientId] = new List<ConnectionState>();
+                InitializeInstance($"Client{clientId}", m_ClientNetworkManagers[clientId], m_ClientConnectionStateSequences[clientId], playerPrefabGO, out m_ClientConnectionManagers[clientId], out m_ClientScopes[clientId]);
+            }
+
+            m_ServerConnectionStateSequence = new List<ConnectionState>();
+            InitializeInstance("Server", m_ServerNetworkManager, m_ServerConnectionStateSequence, playerPrefabGO, out m_ServerConnectionManager, out m_ServerScope);
+
+            base.OnServerAndClientsCreated();
+        }
+
+        protected override IEnumerator OnTearDown()
+        {
+            m_ServerConnectionManager.RequestShutdown();
+
+            yield return new WaitWhile(() => m_ServerNetworkManager.IsListening);
+
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                var clientId = i;
+                yield return new WaitWhile(() => m_ClientNetworkManagers[clientId].IsListening);
+            }
+
+            m_ServerScope.Dispose();
+
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                m_ClientScopes[i].Dispose();
+            }
+
+            foreach (var sceneGameObject in GameObject.FindObjectsOfType<GameObject>())
+            {
+                GameObject.DestroyImmediate(sceneGameObject);
+            }
+            yield return base.OnTearDown();
+        }
+
+        void StartHost()
+        {
+            m_ServerConnectionManager.StartHostIp("server", "127.0.0.1", 9998);
+        }
+
+        IEnumerator ConnectClients()
+        {
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                m_ClientConnectionManagers[i].StartClientIp($"client{i}", "127.0.0.1", 9998);
+            }
+
+            yield return WaitForClientsConnectedOrTimeOut(m_ClientNetworkManagers);
+        }
+
+        void SetUniqueProfilesForEachClient()
+        {
+            for (var i = 0; i < NumberOfClients; i++)
+            {
+                var profileManager = m_ClientScopes[i].Container.Resolve<ProfileManager>();
+                profileManager.Profile = $"Client{i}";
+            }
+        }
+
+        [Test]
+        public void StartHost_Success()
+        {
+            StartHost();
+            Assert.IsTrue(m_ServerNetworkManager.IsHost);
+            var expectedServerConnectionStateSequence = new List<ConnectionState>();
+            expectedServerConnectionStateSequence.Add(m_ServerConnectionManager.m_StartingHost);
+            expectedServerConnectionStateSequence.Add(m_ServerConnectionManager.m_Hosting);
+            Assert.AreEqual(expectedServerConnectionStateSequence, m_ServerConnectionStateSequence);
```

Comment:
Is this test's goal to test internal states themselves or to test the behaviour of the state machine? An issue with testing the internal sequence of a state machine is you're not refactor friendly with this. Usually, when testing state machines, you want to test the inputs and the effects of it, not how it behaves internally. Is it really that important that the state machine went from state 1 to 2 to 3? Usually no as long as the right inputs produced the right outputs.

Reply:
unless you want to test each states individually, but then you'd want to stub a bunch of things and create unit tests for each state and thoroughly test those.
It's the same thinking as if you were testing a method with inputs and return values. You wouldn't care about it's internal state, as long as it gave you the right output.

Reply:
That is a good point, I'll simplify this to only concern myself with the end results instead of the internal states. That way, I could also remove the AssemblyInfo.cs that I added to make internals visible to the tests

Reply:
Unless we want to keep testing the final states at the end of the tests? These are internal, but they affect how inputs are handled from the outside

Reply:
As discussed, testing the state machine is higher prio than individual states.